### ISSUE TITLE
Last release caused reportUnknownVariableType to appear even when typed

### ISSUE THREAD

**Issue by thesadru** [microsoft/pyright#4831]
reportUnknownVariableType now shows up even for annotated types if an unknown variable is used anywhere in the expression. This causes issues with partially untyped libraries.

```py
# library - must be in site-packages to show up
# not type-complete
def function() -> dict:
    return {"hello": "world"}
```
```py
import library

x: dict[str, str] = library.function()  # reportUnknownMemberType (acceptable)
reveal_type(x)  # Type of "x" is "Unknown | dict[str, str]"

print(x["hello"])  # ❌ reportUnknownVariableType
```

**Comment by erictraut** [comment]
Yes, this change was by design. Previously, an unknown type could have snuck in undetected, which means that you would potentially have a type checking "hole" without realizing it. The purpose of the strict-mode `reportUnknownXXX` diagnostics is to report all such conditions, and it wasn't doing so previously. If you have partially untyped libraries, it will be difficult to use these strict-mode checks.

**Comment by thesadru** [comment]
Is `typing.cast` the only alternative in this case? I do not wish to keep unknown variables in my own codebase but AFAIK diagnostics can't be disabled on a per-file basis.

**Comment by erictraut** [comment]
Options include:
1. Writing your own (partial) type stubs to fill in missing information for the library. Minimally, you could create a partial type stub that simply includes a `__getattr__` that returns `Any` for all symbols in that file, or you can selectively add type information as needed.
2. Using `cast` calls.
3. Using `# type: ignore` or `# pyright: ignore` comments to suppress a diagnostic on that line.
4. Using `# pyright: reportUnknownVariableType=false` at the top of a source file to suppress all of the diagnostics of that category within the file.

**Comment by thesadru** [comment]
I've been going with a mixture of 2 and 3 but I didn't know about the possibility of 4, thank you for your help.
